internal/ext/unsafex/unsafex.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package unsafex contains extensions to Go's package unsafe.
+//
+// Importing this package should be treated as equivalent to importing unsafe.
+package unsafex
+
+import "unsafe"
+
+// Size is like [unsafe.Sizeof], but it is a generic function and it returns
+// an int instead of a uintptr (Go does not have types so large they would
+// overflow an int).
+func Size[T any]() int {
+	var zero T
+	return int(unsafe.Sizeof(zero))
+}
+
+// Index is like [unsafe.Add], but it operates on a typed pointer and scales the
+// offset by that type's size, similar to pointer arithmetic in Rust or C.
+//
+// This function has the same safety caveats as [unsafe.Add].
+func Index[T any](p *T, idx int) *T {
+	raw := unsafe.Pointer(p)
+	raw = unsafe.Add(raw, idx*Size[T]())
+	return (*T)(raw)
+}
+
+// Cast performs an unchecked cast of one pointer type to another, through
+// an [unsafe.Pointer].
+func Cast[To, From any](p *From) *To {
+	return (*To)(unsafe.Pointer(p))
+}
+
+// Transmute bit-casts a value of type From to a value of type To.
+//
+// This operation is very dangerous, because it can be used to break package
+// export barriers, read uninitialized memory, and forge pointers in violation
+// of [unsafe.Pointer]'s contract, resulting in memory errors in the GC.
+//
+// Transmute should only be used on types the caller owns, and those types
+// should have identical GC shapes.
+func Transmute[To, From any](p From) To {
```

Comment:
`Cast` and `Transmute` superficially look/sound like different operations (since the choice in words is so different), but they are really the same. Also, `Cast` is not technically necessary since this function can also be used to reinterpret pointer types.

So how about call this one `Cast` and then inline the other function into this one?

Reply:
~Hmm... I don't think I agree. I think we can rename `Transmute` to `Bitcast` (the former is what Rust calls it, the latter is what C++ calls it; could also do `Reinterpret`).~

~I would prefer to keep `Cast` because using `Bitcast` for this is, uh, really scary. It's very easy to mix up `Transmute[*byte](p)` and `Transmute[byte](p)`.~

Thinking about this again I realized Go's lack of type inference helps us here. If I wrote

```go
TakesPointer(Transmute[byte](p))
```

will be an error. This only really hurts us if there is confusion between `*byte` and `**byte` but that's also a problem with `Cast`.

I'm going to rename it to `Bitcast` though. I don't think `Cast` is appropriate, because `Bitcast[float64](int)` is not an i2f conversion, but a bitcast.